File: pages/checkout_page.py
from playwright.sync_api import Page, expect
from .base_page import BasePage


class CheckoutPage(BasePage):
    FIRST_NAME_INPUT = "[data-test='firstName']"
    LAST_NAME_INPUT = "[data-test='lastName']"
    POSTAL_CODE_INPUT = "[data-test='postalCode']"
    CONTINUE_BUTTON = "[data-test='continue']"
    CANCEL_BUTTON = "[data-test='cancel']"
    FINISH_BUTTON = "[data-test='finish']"
    ERROR_MESSAGE = "[data-test='error']"
    SUMMARY_TOTAL = "[data-test='total-label']"
    COMPLETE_HEADER = "[data-test='complete-header']"
    BACK_HOME_BUTTON = "[data-test='back-to-products']"

    def __init__(self, page: Page):
        super().__init__(page)

    def fill_shipping_info(self, first_name: str, last_name: str, postal_code: str):
        self.page.fill(self.FIRST_NAME_INPUT, first_name)
        self.page.fill(self.LAST_NAME_INPUT, last_name)
        self.page.fill(self.POSTAL_CODE_INPUT, postal_code)
        return self

    def click_continue(self):
        self.page.click(self.CONTINUE_BUTTON)
        return self

    def click_finish(self):
        self.page.click(self.FINISH_BUTTON)
        return self

    def click_cancel(self):
        self.page.click(self.CANCEL_BUTTON)
        return self

    def get_error_message(self) -> str:
        return self.page.text_content(self.ERROR_MESSAGE)

    def expect_error_visible(self):
        expect(self.page.locator(self.ERROR_MESSAGE)).to_be_visible()

    def get_total(self) -> str:
        return self.page.text_content(self.SUMMARY_TOTAL)

    def get_complete_header(self) -> str:
        return self.page.text_content(self.COMPLETE_HEADER)

    def expect_order_complete(self):
        expect(self.page.locator(self.COMPLETE_HEADER)).to_have_text(
            "Thank you for your order!"
        )

    def go_back_home(self):
        self.page.click(self.BACK_HOME_BUTTON)
        return self

File: src/pages/CheckoutPage.ts
import { Page, expect } from '@playwright/test';
import { BasePage } from './BasePage';

export class CheckoutPage extends BasePage {
  private readonly firstNameInput = "[data-test='firstName']";
  private readonly lastNameInput = "[data-test='lastName']";
  private readonly postalCodeInput = "[data-test='postalCode']";
  private readonly continueButton = "[data-test='continue']";
  private readonly cancelButton = "[data-test='cancel']";
  private readonly finishButton = "[data-test='finish']";
  private readonly errorMessage = "[data-test='error']";
  private readonly summaryTotal = "[data-test='total-label']";
  private readonly completeHeader = "[data-test='complete-header']";
  private readonly backHomeButton = "[data-test='back-to-products']";

  constructor(page: Page) {
    super(page);
  }

  async fillShippingInfo(firstName: string, lastName: string, postalCode: string): Promise<this> {
    await this.page.fill(this.firstNameInput, firstName);
    await this.page.fill(this.lastNameInput, lastName);
    await this.page.fill(this.postalCodeInput, postalCode);
    await this.pause();
    return this;
  }

  async clickContinue(): Promise<this> {
    await this.page.click(this.continueButton);
    await this.pause();
    return this;
  }

  async clickFinish(): Promise<this> {
    await this.page.click(this.finishButton);
    await this.pause();
    return this;
  }

  async clickCancel(): Promise<this> {
    await this.page.click(this.cancelButton);
    await this.pause();
    return this;
  }

  async getErrorMessage(): Promise<string | null> {
    return this.page.textContent(this.errorMessage);
  }

  async getTotal(): Promise<string | null> {
    return this.page.textContent(this.summaryTotal);
  }

  async expectOrderComplete(): Promise<void> {
    await expect(this.page.locator(this.completeHeader)).toHaveText('Thank you for your order!');
  }

  async expectErrorVisible(): Promise<void> {
    await expect(this.page.locator(this.errorMessage)).toBeVisible();
  }

  async goBackHome(): Promise<this> {
    await this.page.click(this.backHomeButton);
    await this.pause();
    return this;
  }
}
